Jakarta Authorization 1.5

arjantijms opened the pull request:
Signed-off-by: arjantijms <arjan.tijms@gmail.com>

## Specification PR template
When creating a specification project release review, create two PRs with the content divided as follows.

Include the following in PR#1:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [X] Specification PDF in the form of wombat_x.y.pdf
- [X] Specification HTML in the form of wombat_x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [X] For a Release Review, a summary that a Compatible Implementation
      is complete, passes the TCK, and that the TCK includes sufficient
      coverage of the specification. The TCK users guide MUST include
      the instructions to run the compatible implementations used to
      validate the release.  Instructions MAY be by reference.
- [X] The URL of the OSSRH staging repository for the api, javadoc:
      https://oss.sonatype.org/content/repositories/staging/jakarta/authorization/jakarta.authorization-api/1.5.0/
- [X] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee8-eftl/promoted/eclipse-authorization-tck-1.5.0.zip
- [x] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/jacc/issues/46

dblevins commented:
Spec status is Draft, but should be Final or Final Release.

arjantijms commented:
> Spec status is Draft, but should be Final or Final Release.
> Also version number is still SNAPSHOT

Thanks, I'll update it.

joakime commented on authorization/1.6/_index.md:
```diff
@@ -0,0 +1,37 @@
+---
+Jakarta Authorization defines a low-level SPI for authorization modules, which are repositories of permissions 
+facilitating subject based security by determining whether a given subject has a given permission, and algorithms
+to transform security constraints for specific containers (such as Jakarta- Servlet or Enterprise Beans) into 
```
remove dash, `s/Jakarta- Servlet/Jakarta Servlet`

arjantijms replied:
The idea was to shorten it, so it would be instead "Jakarta Servlet or Jakarta Enterprise Beans", but I'll update it. Thx

joakime replied:
Ah, I see what you were trying to do now.
Didn't come across like that on first read.
In this case I see no harm in being verbose.

arjantijms replied:
Okay, I updated it everywhere, thank again!

bshannon commented on authorization/1.6/_index.md:
```diff
@@ -0,0 +1,38 @@
+* [Jakarta Authorization 1.6 Specification Document](./authorization-spec-1.6.pdf) (PDF)
+* [Jakarta Authorization 1.6 Specification Document](./authorization-spec-1.6.html) (HTML)
+* [Jakarta Authorization 1.6 Javadoc](./apidocs)
+* [Jakarta Authorization 1.6 TCK](https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee8-eftl/promoted/eclipse-authorization-tck-1.5.0.zip)
```
This needs to be the URL of the final signed version, which will be at:
https://download.eclipse.org/jakartaee/authorization/1.5/eclipse-authorization-tck-1.5.0.zip

bshannon commented on authorization/1.6/_index.md:
```diff
@@ -0,0 +1,38 @@
+---
+title: "Jakarta Authorization 1.6"
```
The version of JACC in Java EE 8 is 1.5, I don't know why this is 1.6 but I believe it should all be 1.5.

arjantijms replied:
@bshannon the last version released by Oracle in Maven central was unfortunately set at 1.6:

https://mvnrepository.com/artifact/javax.security.jacc/javax.security.jacc-api/1.6

I discussed this last time with Dmitry for the GlassFish 5.1 release and there we decided to release it as 1.6:

https://mvnrepository.com/artifact/jakarta.security.jacc/jakarta.security.jacc-api

arjantijms commented:
@bshannon regarding the 1.5 version:

The last version released by Oracle in Maven central was unfortunately set at 1.6:

https://mvnrepository.com/artifact/javax.security.jacc/javax.security.jacc-api/1.6

I discussed this last time with Dmitry for the GlassFish 5.1 release and there we decided to release it as 1.6:

https://mvnrepository.com/artifact/jakarta.security.jacc/jakarta.security.jacc-api

starksm64 commented:
# Spec Review Checklist

1. Spec PR
  - [x] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [x] Directory of form {spec}/x.y
  - [x] PDF of form {spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [x] HTML of form {spec}-spec-x.y.html ("-spec" preferred but not required)
  - [x] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [x] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [x] Staging repository link of the form https://oss.sonatype.org/conte/repositories/staging/jakarta/{spec}/jakarta-{spec}-api/x.y.z/
  - [x] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [x] Second PR for just apidocs

2. _index.md
  - [x] Link to spec pdf
  - [x] Link to spec html
  - [x] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip
  - [x] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [x] Name of and link to at least one Compatible Implementation

3. javadocs
  - [x] Footer contains Eclipse copyright and link to license
  - [x] ESFL license is included, usually as doc-files/speclicense.html
  - [x] no META-INF directory in PR
  - [x] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [x] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [x] Correct Eclipse copyright line
  - [x] No DRAFT or SNAPSHOT
  - [x] Correct Logo

5. Spec HTML
  - [x] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [x] EFTL license file, preferably named LICENSE.md
  - [x] User's Guide (or equivalent documentation)
  - [x] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [x] Software requirements listed
  - [x] Installation and configuration described
  - [x] How to run tests
  - [x] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
81373a045b7bb187989eafbf6c4be5b6f7f28d25cdbdae03e2fcf343889f77f2
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors

kwsutter commented:
Looks like we're still waiting for this to be changed over to version 1.5 across the board...

arjantijms commented:
> Looks like we're still waiting for this to be changed over to version 1.5 across the board...

I'm still not entirely sure what the final-final decision is on this. I'll comment on the list to ask.

starksm64 commented:
That is the last thing to resolve. I thought was we wanted to do was use 1.5.0 with 
Maven coordinates of jakarta.authorization:jakarta.authorization-api:1.5.0
and keep the spec at version 1.5, and rerun the standalone TCK
against the updated API jar file.

bshannon commented:
> That is the last thing to resolve. I thought was we wanted to do was use 1.5.0 with
> Maven coordinates of jakarta.authorization:jakarta.authorization-api:1.5.0
> and keep the spec at version 1.5, and rerun the standalone TCK
> against the updated API jar file.

I agree.

arjantijms commented:
> I agree.

Perfect, then that's what it's going to be. I'll do the change over and re-test.

arjantijms commented:
@bshannon @starksm64 I created a bug to request the new Maven coordinates here: 

https://bugs.eclipse.org/bugs/show_bug.cgi?id=550199

starksm64 commented:
I'm not seeing an updated staging release. Was one built?

arjantijms commented:
> I'm not seeing an updated staging release. Was one built?

About to, ran into some issues with the Servlet docs, but they are fixed now.

arjantijms commented:
@starksm64 It's there now:

* https://oss.sonatype.org/content/repositories/staging/jakarta/authorization/jakarta.authorization-api/
* https://github.com/eclipse-ee4j/jacc/releases/tag/1.5.0-RELEASE

starksm64 commented:
Ok, ack that I can see it now.

arjantijms commented:
Ran into a small obstacle, and that's that in OSGi the package names are versioned. Since we changed the version of the jar from 1.6 to 1.5, OSGi now complains in GlassFish 5.1 when you just replace the jar:

```
missing requirement [org.glassfish.main.web.glue [273](R 273.0)] osgi.wiring.package; 
    (&(osgi.wiring.package=com.sun.logging.enterprise.system.container.web)(version>=5.1.0)(!(version>=6.0.0))) 
         [caused by: Unable to resolve org.glassfish.main.web.war-util [269](R 269.0): missing requirement 
              [org.glassfish.main.web.war-util [269](R 269.0)] osgi.wiring.package; 
                   (&(osgi.wiring.package=com.sun.enterprise.security.perms)(version>=5.1.0)(!(version>=6.0.0))) 
                    [caused by: Unable to resolve org.glassfish.main.security.ee [248](R 248.0): 
                    missing requirement [org.glassfish.main.security.ee [248](R 248.0)] 
                    osgi.wiring.package; (&(osgi.wiring.package=javax.security.jacc)(version>=1.6.0)(!(version>=2.0.0)))]]] 
```

I have to think a little about the best way to solve this. One option could be to update the manifest during the test to 1.6. Since it's not part of the spec or api I think that would be allowed. Alternatively GF can be build on the fly with the new dependency.

starksm64 commented:
> I have to think a little about the best way to solve this. One option could be to update the manifest during the test to 1.6. Since it's not part of the spec or api I think that would be allowed. Alternatively GF can be build on the fly with the new dependency.

I think building GF on the fly is most consistent with what we have been doing elsewhere.

arjantijms commented:
@starksm64 
> I think building GF on the fly is most consistent with what we have been doing elsewhere.

I'll do that, although it hasn't been exactly without (small) issues, such as https://github.com/eclipse-ee4j/glassfish/issues/22782

Do you know which other projects are building GF on the fly?

starksm64 commented:
By on the fly I mean they are creating a dist of glassfish with the updated apis, usually as a separate job and then using that as input into the tck run.

arjantijms commented:
@starksm64 Thanks for the example, but that's what I'm normally doing already ;)

In this case however it's more involved, and I actually need to checkout from source, update the coordinates and version via SED and then literally build it using `mvn`.

This is because the version number changed, and all the code in GF has OSGi statements for package `javax.security.jacc` at version 1.6 or higher. So just replacing doesn't work, as it will find a `javax.security.jacc` package at version 1.5.

It's due to the way OSGi works, and is a GF specific peculiarity.

But it's okay, I've got most everything running now.

arjantijms commented:
@starksm64 @bshannon re-ran the TCK against a GF build from source; ["5.1.0" tag](https://github.com/eclipse-ee4j/glassfish/tree/5.1.0) using `sed` to update Jakarta Authorization to 1.5.0 at the new maven coordinates `jakarta.authorization:jakarta.authorization-api:1.5.0`.

Additionally updated the script to also execute the signature tests, which now pass (there was one failure, which I fixed).

Updated Javadoc: https://github.com/jakartaee/specifications/pull/55/commits/01bf0be15e001795f908b0635b8a67849002e811
Updated TCK results with the extra test here: https://github.com/eclipse-ee4j/glassfish/pull/22826

This was a bit of extra work altogether, but we should now be good to go.

bshannon commented:
GlassFish needs to be updated to build from source with the updated dependency.
What the script above is doing is effectively "binary patching" GlassFish.  That may
be sufficient for running the TCK against GlassFish, but it doesn't allow us to release
a new version of GlassFish with the updated dependency.

arjantijms commented:
>GlassFish needs to be updated to build from source with the updated dependency.

Indeed, that's what I have been doing. Pending the read access request, I'll post the script I'm using here:

```groovy
#!/usr/bin/env groovy

node {
    
    def GF_API_JAR_NAME = "jakarta.security.jacc-api.jar"
    
    def API_JAR_NAME = "jakarta.authorization-api.jar"
    def DOWNLOAD_API_JAR_NAME = "jakarta.authorization-api-1.5.0.jar"
        
    def API_URL = "https://oss.sonatype.org/content/repositories/staging/jakarta/authorization/jakarta.authorization-api/1.5.0/${DOWNLOAD_API_JAR_NAME}"
    def TCK_BUNDLE_URL = "http://download.eclipse.org/ee4j/jakartaee-tck/jakartaee8-eftl/promoted/eclipse-authorization-tck-1.5.0.zip"
    
    
    // TCK properties
    env.deliverabledir="authorization"
    env.TCK_HOME="${env.WORKSPACE}"
    env.TS_HOME="${env.WORKSPACE}/${env.deliverabledir}-tck"
    env.javaee_home="${env.WORKSPACE}/glassfish5"
    
    env.ANT_VERSION='1.9.13'
    env.TOOLS_PREFIX='/opt/tools'
    env.JAVA_PREFIX="${TOOLS_PREFIX}/java/oracle"
    env.MVN_HOME="${TOOLS_PREFIX}/apache-maven/latest"
    env.JAVA_HOME="${JAVA_PREFIX}/jdk-8/latest"
    env.PATH="${MVN_HOME}/bin:${JAVA_HOME}/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    env.ANT_HOME="${TS_HOME}/tools/ant"
    env.PATH="${ANT_HOME}/bin:${PATH}"
    env.ANT_OPTS="-Djavax.xml.accessExternalSchema=all"
    
    stage("Init") {
        // https://go.cloudbees.com/docs/cloudbees-documentation/cjoc-user-guide/index.html#cluster-copy-artifacts
        dir ("download") {
            // If any artefacts from other jobs need to be used, they can be placed here
            // copyRemoteArtifacts(from: "${GF_BUILD_JOB}")
            // copyRemoteArtifacts(from: "${TS_JTE_BUILD_JOB}")
        }
        sh '''#!/bin/bash -x
        
            pwd
            mkdir download
            mkdir -p jacctckreport/jacctck
            ls -altrh
            
           '''
    }
    
    stage('Grab GF src') {
           dir ("checkout") {
            echo '*#*#*#*#*#*#*#*#*#*#*#*#  Checking out GlassFish  *#*#*#*#*#*#*#*#*#*#*#*#*#*#*#'
            checkout changelog: false, poll: false, scm: [$class: 'GitSCM',
                branches: [[name: "refs/tags/5.1.0"]],
                userRemoteConfigs: [[url: "https://github.com/eclipse-ee4j/glassfish.git"]]]
            echo '*#*#*#*#*#*#*#*#*#*#*#*#  Checked out GlassFish  *#*#*#*#*#*#*#*#*#*#*#*#*#*#*#'
           }
    
        sh '''#!/bin/bash -x
            pwd
            ls -altrh
            cd checkout
            grep -rl jakarta.security.jacc . | tee replace.txt | xargs sed -i "s#jakarta.security.jacc#jakarta.authorization#g"
            sed -i "s#1.6.1#1.5.0#g" appserver/pom.xml
            
            echo ### replacements in GlassFish source:
            cat replace.txt
            rm replace.txt
            
            echo ################### START appserver pom.xml
            cat appserver/pom.xml
            echo ################### END appserver pom.xml
            
            
        '''
    }
    
    stage("Build Glassfish") {
    	
    	echo '*#*#*#*#*#*#*#*#*#*#*#*#  Running build  *#*#*#*#*#*#*#*#*#*#*#*#*#*#*#'
        sh '''#!/bin/bash -ex
        
            cd ${WORKSPACE}/checkout
            
            MVN_EXTRA='-U --show-version --batch-mode -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn -Pstaging'
            
            mvn $MVN_EXTRA
            
           '''
           echo '*#*#*#*#*#*#*#*#*#*#*#*#  Ran build  *#*#*#*#*#*#*#*#*#*#*#*#*#*#*#'
    }
    
    stage("Grab Glassfish") {
    	sh '''#!/bin/bash -ex
        
            cd ${WORKSPACE}/download
            cp ${WORKSPACE}/checkout/appserver/distributions/glassfish/target/glassfish.zip .
            
            cd ${WORKSPACE}
            unzip -q ${WORKSPACE}/download/glassfish.zip
            
           '''
    }

    stage("Grab TCK") {
        env.TCK_BUNDLE_URL = "${TCK_BUNDLE_URL}"
        sh '''#!/bin/bash -ex
            cd ${WORKSPACE}/download 
            wget -q ${TCK_BUNDLE_URL} -O ${deliverabledir}tck.zip
            
            cd ${WORKSPACE}
            unzip ${WORKSPACE}/download/${deliverabledir}tck.zip
           '''
    }
   
    stage ("Grab ANT") {
        sh '''#!/bin/bash -ex
        
            cd ${WORKSPACE}/download
            wget -q http://mirror.koddos.net/apache/ant/binaries/apache-ant-${ANT_VERSION}-bin.tar.gz -O ant.tar.gz
            tar xfz ant.tar.gz
            mkdir -p ${ANT_HOME} && cp -a ${WORKSPACE}/download/apache-ant-${ANT_VERSION}/. ${ANT_HOME}
           
           '''
    }

    stage ("Configure ts.jte") {
       
        sh '''#!/bin/bash -ex
           
            cd ${TS_HOME}/bin/
            
            sed -i 's#orb.port=.*#orb.port=3699#g' ts.jte
            sed -i 's#javaee.level=.*#javaee.level=full#g' ts.jte
            sed -i "s#jacc.home=.*#jacc.home=$TCK_HOME/glassfish5/glassfish#g" ts.jte
            sed -i 's#jacc.host=.*#jacc.host=localhost#g' ts.jte
            sed -i "s#^report.dir=.*#report.dir=$TCK_HOME/jacctckreport/jacctck#g" ts.jte
            sed -i "s#^work.dir=.*#work.dir=$TCK_HOME/jacctckwork/jacctck#g" ts.jte

            sed -i "s#jakarta.security.jacc#jakarta.authorization#g" ts.jte

            export CONTENT='<property name="all.test.dir" value="com/sun/ts/tests/jacc/,com/sun/ts/tests/signaturetest/jacc,com/sun/ts/tests/common/vehicle/" />'
            sed -i "\\#</project># s#.*#${CONTENT}\\n&#" build.xml
            
                       
           '''
    }
    
    stage("Configure TCK") {
        sh '''#!/bin/bash -ex
            cd $TCK_HOME/glassfish5/glassfish/bin
            ./asadmin start-domain
           
            cd ${TS_HOME}/bin
            ant config.vi
            
            cd $TS_HOME/bin
            ant enable.jacc
            
           '''
    }
   
    stage ("Run TCK tests") {
         sh '''#!/bin/bash -x
         
         cd $javaee_home/bin
         ./asadmin stop-domain
         
         if [ $retval != 0 ]; then
            echo "Pending process to be killed:"
            ps -eaf | grep "com.sun.enterprise.admin.cli.AdminMain" | grep -v "grep" | grep -v "nohup" 
            for i in `ps -eaf | grep "com.sun.enterprise.admin.cli.AdminMain" | grep -v "grep" | grep -v "nohup" | tr -s " " | cut -d" " -f2`
            do
                echo "[killJava.sh] kill -9 $i"
                 kill $i
             done
         fi
         
         ./asadmin start-domain
         
         cd $TS_HOME/src/com/sun/ts/tests/
         ant deploy run.all | tee -a ${TS_HOME}/bin/run.log
           
           '''       
    } 
    
    stage ("Create summary.txt, API, and run.log artifacts") {
        env.API_JAR_NAME = "${API_JAR_NAME}"
        sh '''#!/bin/bash -ex
            cd ${TS_HOME}/bin
            # cat run-ejb.log run-web.log > run.log
            cat run.log | sed -e '1,/Completed running/d' > summary.txt
            PASSED_COUNT=`head -1 summary.txt | tail -1 | sed 's/.*=\\s\\(.*\\)/\\1/'`
            FAILED_COUNT=`head -2 summary.txt | tail -1 | sed 's/.*=\\s\\(.*\\)/\\1/'`
            ERROR_COUNT=`head -3 summary.txt | tail -1 | sed 's/.*=\\s\\(.*\\)/\\1/'`
            
            echo ERROR_COUNT=${ERROR_COUNT}
            echo FAILED_COUNT=${FAILED_COUNT}
            echo PASSED_COUNT=${PASSED_COUNT}   
            
            SHA256_API=`openssl dgst -sha256 ${TCK_HOME}/glassfish5/glassfish/modules/${API_JAR_NAME}`
            SHA256_TCK=`openssl dgst -sha256 ${WORKSPACE}/download/${deliverabledir}tck.zip`
         
            echo SHA256_API=${SHA256_API} | tee -a summary.txt
            echo SHA256_TCK=${SHA256_TCK}  | tee -a summary.txt
           '''

        archiveArtifacts artifacts: "${env.deliverabledir}-tck/bin/summary.txt", fingerprint: true
        archiveArtifacts artifacts: "${env.deliverabledir}-tck/bin/run.log", fingerprint: true
        archiveArtifacts artifacts: "glassfish5/glassfish/modules/${API_JAR_NAME}", fingerprint: true
        archiveArtifacts artifacts: "${env.deliverabledir}-tck/bin/ts.jte", fingerprint: true
        
    }
}
```

arjantijms commented:
@bshannon 
> it doesn't allow us to release a new version of GlassFish with the updated dependency.

PR for GlassFish with the updated dependency for Jakarta Authorization 1.5: https://github.com/eclipse-ee4j/glassfish/pull/22827

bshannon commented:
* On ballot completion
 - [ ] The designated specification committee representative adds this final checklist to the main PR.
 - [ ] The specification committee representative merges the specification and apidocs PRs.
 - [ ] The designated specification committee representative calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job.
 - [ ] The specification project member who created the staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
 - [ ] The EMO updates the specification page with the ballot results.
This list goes on the committed spec index page.
 - [ ] The specification project team should go through the merged spec website page to verify all the links are valid.
 - [ ] The specification project team should approve the compatibility request.
 - [ ] The compatible implementation project/vendor MUST send an email tck@eclipse.org for approval of the compatible implementation for trademark usage.
 - [ ] The specification project team should merge any final release branch as appropriate for the branch management for the project.

bshannon commented:
@arjantijms Please let me know when you've completed your items in the checklist above.  Thanks.

arjantijms commented:
@bshannon 

The specification API jars have been promoted to Maven central, see:

https://repo1.maven.org/maven2/jakarta/authorization/jakarta.authorization-api/1.5.0/

bshannon commented:
@arjantijms Thanks.  Let me know when the remaining items have been completed.

arjantijms commented:
@bshannon due to the employment transition I still don't have full acces back. For instance, I can't add labels to the CR issue here: https://github.com/eclipse-ee4j/jacc/issues/46

arjantijms commented:
Final release branch merge PR: https://github.com/eclipse-ee4j/jacc/pull/51

bshannon commented:
> For instance, I can't add labels to the CR issue here: [eclipse-ee4j/jacc#46](https://github.com/eclipse-ee4j/jacc/issues/46)

Do you believe it should be marked accepted and closed?  Add a comment to the issue.

arjantijms commented:
Added comment: https://github.com/eclipse-ee4j/jacc/issues/46#issuecomment-529751151

arjantijms commented:
Access is coming back. Accepted and closed CR issue: https://github.com/eclipse-ee4j/jacc/issues/46

arjantijms commented:
* Links checked
* Mail sent
